### PR TITLE
fix(core): do not drop signals added during BatchExecutor flush

### LINKED ISSUE #2254: BatchExecutor flush silently discards signals pushed during the flush
## Parent

- Session replay delivery reliability epic — `grafana/faro-web-sdk#2241`

## What happens

`BatchExecutor.flush()` sends the buffered signal groups synchronously (`itemGroups.forEach(this.sendFn)`) and only afterwards resets the buffer with `this.signalBuffer = []` (`packages/core/src/transports/batchExecutor.ts`). Any signal pushed through `addItem` while `sendFn` is still executing lands in the buffer between those two steps and is silently discarded — no log, no error, no retry.

Re-entrant pushes during a flush are not exotic. The session instrumentation registers a transport `beforeSend` hook that runs `updateSession()` per item during the send; a session rotation there calls `faro.api.setSession`, metas listeners fire synchronously, and any listener that emits telemetry in response pushes items mid-flush. The replay instrumentation is a concrete victim: rotation stops and restarts recording, and the fresh Meta, FullSnapshot, and `faro.session_recording.started` events it emits are wiped by the buffer reset. The post-rotation recording then starts with orphan incremental events and stays unplayable until the next 5-minute checkout.

The itemLimit-triggered flush (inside `addItem`) and the `visibilitychange: hidden` flush have the same shape, so the loss window exists on every flush path.

## What to fix

`flush()` must not discard items added during the send. Snapshot-and-swap the buffer before iterating (take the current buffer, assign a fresh empty one, then group and send the snapshot), so re-entrant pushes accumulate for the next flush instead of vanishing.

## Acceptance criteria

- [ ] A signal pushed while `flush()` is sending is delivered in a later flush, never discarded.
- [ ] Items sent in one flush keep their order; re-entrant items do not join the in-flight batch.
- [ ] A regression test drives a re-entrant `addItem` from inside a mocked send function and asserts the item is delivered.

## Blocked by

None — can start immediately.

## Further notes

- Found while specifying `grafana/faro-web-sdk#2242`. That issue defers the replay rotation restart out of the flush call stack as a local workaround; this issue is the general fix benefiting all signal types.

### PR DESCRIPTION
**Pull request by RealBhupesh**
## Summary

`BatchExecutor.flush()` grouped and sent `this.signalBuffer`, then assigned `this.signalBuffer = []`. `sendFn` is synchronous, so `beforeSend` / `setSession` can `addItem` during send. Those items were wiped.

## Fix

Snapshot-and-swap before iterating: capture the buffer, reset it, then group/send the snapshot. Re-entrant items land on the new buffer and go out on the next flush.

## Tests

Regression in `batchExecutor.test.ts`: mock `sendFn` that calls `addItem`; the extra item is delivered on the following flush.

Fixes #2254

### CODE REVIEW

File: packages/core/src/transports/batchExecutor.ts [hunk 1]
```diff
@@ -81,8 +81,15 @@ export class BatchExecutor {
...
-    const itemGroups = this.groupItems(this.signalBuffer);
-    itemGroups.forEach(this.sendFn);
+    const snapshot = this.signalBuffer;
     this.signalBuffer = [];
+
+    try {
+      const itemGroups = this.groupItems(snapshot);
+      itemGroups.forEach(this.sendFn);
+    } catch (error) {
+      this.signalBuffer = [...snapshot, ...this.signalBuffer];
```
Comment (kpelelis) [inline packages/core/src/transports/batchExecutor.ts:91]: This effectively overrides the limit of the buffer opening up for a memory overflow if the send function is constantly failing.

The optimal solution here would be to have a retry/exponential backoff mechanism but in order not to blow up the scope of this PR, trimming the buffer to the limit could work as an incomplete alternative.

File: packages/core/src/transports/batchExecutor.test.ts [hunk 2]
```diff
@@ -189,5 +189,86 @@ describe('BatchExecutor', () => {
...
+      expect(mockSendFunction).toHaveBeenCalledTimes(1);
+      expect(mockSendFunction.mock.calls[0]![0]).toEqual([item1]);
+
+      jest.advanceTimersByTime(1000);
+
+      expect(mockSendFunction).toHaveBeenCalledTimes(2);
+      expect(mockSendFunction.mock.calls[1]![0]).toEqual([item2]);
+    });
+
+    it('retries buffered signals after synchronous dispatch fails', () => {
```
Comment (kpelelis) [inline packages/core/src/transports/batchExecutor.test.ts:222]: I would add a corner case where the buffer is almost full or full to its limit and after an error happens, there are signals in the buffer again that would overflow the limit

**Review by kpelelis — CHANGES_REQUESTED** [review]
Hey @RealBhupesh this is a very nice find and I am glad you are tackling it. In the lack of a real retry/exponential back off mechanism this is a great alternative and I would be glad to merge it. There is an edge case when running close to the limits of the buffer that could possibly result in undefined behaviour. See my inline comments for the actual issue

Other than that LGTM

File: packages/core/src/transports/batchExecutor.ts (hunk 1 above)
Comment (RealBhupesh) [reply packages/core/src/transports/batchExecutor.ts:91]: Thanks for the review @kpelelis — agreed on the edge case.

After a failed flush we now call `trimBufferToLimit()` so re-entrant `addItem` calls cannot push the buffer past `itemLimit`. It keeps the oldest items (`slice(0, itemLimit)`) so the failed batch is still retried first.

Added a regression test for a full buffer (`itemLimit: 2`) where the send throws after adding a re-entrant item, and verified the buffer stays at length 2 across repeated timer flushes.

Pushed in 5a470526.

File: packages/core/src/transports/batchExecutor.test.ts (hunk 2 above)
Comment (RealBhupesh) [reply packages/core/src/transports/batchExecutor.test.ts:222]: Added `trims the buffer to the item limit after a failed flush when the buffer is already full` in `batchExecutor.test.ts` — covers the full-buffer + failed send + re-entrant `addItem` path and asserts the buffer stays at `itemLimit` across repeated timer flushes.

**Comment by kpelelis** [comment]
@RealBhupesh, LGTM but you need to fix cursoragent so that the CLA is signed

**Comment by RealBhupesh** [comment]
Done. Fixed it!

**Review by kpelelis — APPROVED** [review]

**Comment by martinkuba** [comment]
@RealBhupesh The only thing pending is that we require commits to be signed. Can you please setup commit signing and rebase? Thanks!
